Lotsafixes

rowleya opened the pull request:
Lots of small changes that make things work... includes:
 - New SCAMP binary that ensures that Ethernet chips always work
 - Updated SCAMP configuration value to make the boot more reliable
 - Increase of general timeouts
 - Add some wait time when a retry-able error code is received
 - Reorder some of the SCP pipeline code to work better

andrewgait commented:
I don't have any issues with merging this; it certainly helps a great deal with running larger jobs on bigger machines.  It would be good if at least one other person checked this through though...

alan-stokes commented:
cant find the lines on github. so going to put here. 
trasnciever.py line 359. why make then append, vs what it was before where append was within the make?
line 372. magic number. 
line 944, 949 and 954. redundant brackets. flagged by pycharm
for line 1020 line 70 in version_info.py nmeeds to say tuple of 3 ints, not float
line 1142 why the return? 
line 1431 2 magic numbers.
line 1492 magic number
line 2329 random return.
line 2837 local_port is needed str() as the exception expects a string

proceses. Those magic numbers are being changed but are the same. local area for thems omewhere?

scp_request_pipeline line 210: redundant brackets

apart from these. yep would rubber.

dkfellows commented on spinnman/connections/scp_request_pipeline.py:
```diff
@@ -56,7 +57,7 @@ def __init__(self, connection, n_channels=1,
             to wait for before continuing sending requests. If None, this will\
             be determined automatically
-        :param n_retries: The number of times to resend any packet for any\
+        :param n_retries: The number of times to re-send any packet for any\
```
“resend” is a word

dkfellows commented on spinnman/connections/scp_request_pipeline.py:
```diff
@@ -154,7 +143,21 @@ def send_request(self, request, callback, error_callback):
+        self._retries[sequence] = self._n_retries
+        self._callbacks[sequence] = callback
+        self._error_callbacks[sequence] = error_callback
+        self._retry_reason[sequence] = list()
```
Would it have been better to make `self._retry_reason` be originally a `defaultdict(list)`?

dkfellows commented on spinnman/connections/scp_request_pipeline.py:
```diff
@@ -200,16 +203,19 @@ def _single_retrieve(self, timeout):
+            # If the response can be retried, retry it
+            if (result in RETRY_CODES):
+                try:
+                    time.sleep(0.1)
```
Do you want to sleep before _every_ send? Can we make this be so that it only happens on a retry?

rowleya replied:
This only happens when the result is in the retry codes, so the first send won't go through this.

dkfellows commented on spinnman/transceiver.py:
```diff
@@ -2083,7 +2083,7 @@ def wait_for_cores_to_be_in_state(
 
         # If we haven't reached the final state, do a final full check
-        if processors_ready != len(all_core_subsets):
+        if processors_ready < len(all_core_subsets):
```
What happens when the number of processors ready is greater than the number of cores?

rowleya replied:
This is a bit of a horrid fix in reality, as it is fixing the issue that all_core_subsets might only be a subset of the app_id that you are querying (because of changes in FEC).  If there are more processors ready than I need, these are probably just other processors.  This needs fixing in other ways, but this is reflected in SpiNNakerManchester/SpiNNFrontEndCommon#282.

dkfellows replied:
OK. Well, not OK but whatever.

rowleya commented:
> cant find the lines on github. so going to put here.

If it isn't on github in the files section, I didn't make the change...  If you must do the comparison with PyCharm, please make sure you merge master in to the branch first, as otherwise it will see "changes" that are out of date.

>trasnciever.py line 359. why make then append, vs what it was before where append was within the make?
> line 372. magic number.
> line 944, 949 and 954. redundant brackets. flagged by pycharm
> for line 1020 line 70 in version_info.py nmeeds to say tuple of 3 ints, not float
> line 1142 why the return?
> line 1431 2 magic numbers.
> line 1492 magic number
> line 2329 random return.
> line 2837 local_port is needed str() as the exception expects a string

All valid points, but nothing to do with this PR.

> proceses. Those magic numbers are being changed but are the same. local area for thems omewhere?

Valid point, fixed.

> scp_request_pipeline line 210: redundant brackets

Valid point, fixed.

dkfellows commented on spinnman/connections/scp_request_pipeline.py:
```diff
@@ -224,69 +231,61 @@ def _single_retrieve(self, timeout):
 
-    def _resend(self, seq, request_sent):
+    def _resend(self, seq, request_sent, reason):
         if self._retries[seq] > 0:
```
I notice that we're still using inverted test logic here. It's better to write “if bad-condition: raise exception;<newline>good-thing” instead of “if not bad-condition: good-thing else: raise exception”.
